core/src/apps/zcash/orchard/crypto/sinsemilla.py

```diff
@@ -0,0 +1,59 @@
+"""
+Implementation of Sinsemilla hash function.
+see: https://zips.z.cash/protocol/protocol.pdf#concretesinsemillahash
+"""
+
+from micropython import const
+from typing import TYPE_CHECKING
+
+from trezor.crypto.pallas import group_hash
+from trezor.utils import ensure
+
+from .utils import i2leosp, lebs2ip
+
+if TYPE_CHECKING:
+    from trezor.crypto.pallas import Point
+
+    pass  # sinsemilla.i
+
+
+K = const(10)
+
+
+def iadd(a: Point, b: Point) -> Point:
+    """Incomplete addition."""
+    ensure(not a.is_identity())
+    ensure(not b.is_identity())
+    ensure(a != b)
+    ensure(a != -b)
+    return a + b
+
+
+class Sinsemilla:
+    def __init__(self, acc):
+        self.buffer = []
+        self.acc = acc
+
+    @staticmethod
```

Comment:
all usages of `@staticmethod` in the zcash code should in fact be `@classmethod`s of return type `Self` and return `cls(...)`

Reply:
0299f677207ba93bbac61b533df172f8889f8c81